Fix MotionEye config flow

ZuluWhiskey opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Currently, `Stream URL template` is required when it shouldn't be.
Users may want to enable `Overwrite unrecognized webhooks` or `Configure motionEye webhooks to report events to Home Assistant` but not configure a `Stream URL template`.

This change is to change this configuration option from `Required` to `Optional`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: I have searched and could see no open relevant issues.
- This PR is related to issue: As above.
- Link to documentation pull request: Documentation doesn't currently note this as mandatory so looks like a mistake

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

**Apologies, this is my first PR into core, I don't have the ability to carry out tests using Tox etc., however I have made this change locally and is has resolved the issue. It is a single line change so I hope an exception can be made.**

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ZuluWhiskey committed:
Update config_flow.py

Change the Stream URL template to optional

homeassistant commented:
Hi @ZuluWhiskey, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#64360).

Once you do that we will be able to review and accept this pull request.

Thanks!

bdraco reviewed (changes requested):
Thanks.  Please add a test for this so we can ensure it doesn't regress in the future 👍

ZuluWhiskey commented:
> Thanks.  Please add a test for this so we can ensure it doesn't regress in the future 👍 

Apologies, as I'm going to sound like an idiot here - I've set up the devcontainer on my machine, looking at the documentation it says to run the `tox` command. How can I then show you those results?

bdraco commented:
> > Thanks.  Please add a test for this so we can ensure it doesn't regress in the future 👍
> 
> Apologies, as I'm going to sound like an idiot here - I've set up the devcontainer on my machine, looking at the documentation it says to run the `tox` command. How can I then show you those results?

Commit the test, and push. The CI will run the tests and show the results

ZuluWhiskey commented:
> > > Thanks.  Please add a test for this so we can ensure it doesn't regress in the future 👍
> 
> > 
> 
> > Apologies, as I'm going to sound like an idiot here - I've set up the devcontainer on my machine, looking at the documentation it says to run the `tox` command. How can I then show you those results?
> 
> 
> 
> Commit the test, and push. The CI will run the tests and show the results

Thank you for the tip, I've been trying to sort this on and off for the last couple of days but I'm having no luck. I've done the following:
- I followed the steps in [the documentation](https://developers.home-assistant.io/docs/development_environment/) for VSCode and devcontainer - Using my fork (https://github.com/ZuluWhiskey/core)
- I did a `git fetch` and then switched branch from `dev` to `patch-1`
- Ran the lazytox script but it didn't find any files and recommended I run `tox`
- Ran `tox` but didn't get a successful confirmation as there were a number of failures that (to me) seem completely unrelated to the minor change I made on this one file for this one component

Unfortunately I'm really out of my depth here so I would really appreciate any advice or tips on what I might be doing wrong here? I'm keen to help but understand if this is too much hassle!

DurgNomis-drol commented:
@ZuluWhiskey You need to fix [this](https://github.com/home-assistant/core/blob/d56ab165560b9296f7703ebc061869246e152123/tests/components/motioneye/test_config_flow.py#L468) file and add a test in there. In short, a python test is a function that checks if the code you have written functions as expected. 😃 Hope this helps

ZuluWhiskey committed:
Update test_config_flow.py

Removed URL from the tests so that it is reflective of the Stream URL field being optional

ZuluWhiskey commented:
> @ZuluWhiskey You need to fix [this](https://github.com/home-assistant/core/blob/d56ab165560b9296f7703ebc061869246e152123/tests/components/motioneye/test_config_flow.py#L468) file and add a test in there. In short, a python test is a function that checks if the code you have written functions as expected. 😃 Hope this helps

Thank you so much, shows that I really have no idea what I am doing here!

I have added a commit to that test file now whereby I've removed the http string in the advanced options test - Would that be suffice here or would we need to test both scenarios? ie. one test with a URL populated and another test without the URL being populated? I'm happy to do so if that would be the required option

DurgNomis-drol commented on homeassistant/components/motioneye/config_flow.py:
```diff
@@ -225,7 +225,7 @@ async def async_step_init(
+                    vol.Optional(
                         CONF_STREAM_URL_TEMPLATE,
                         default=self._config_entry.options.get(
                             CONF_STREAM_URL_TEMPLATE,
```
Line below should be `None` instead of an empty string

DurgNomis-drol replied:
You need to change the line below the one I have tagged to `None`. I cannot tag the lines as Github only allows me to tag 3 lines from the thing you have changed 😃

ZuluWhiskey replied:
> You need to change the line below the one I have tagged to `None`. Right now it is an empty string `""`. I cannot tag the lines as Github only allows me to tag 3 lines from the thing you have changed 😃

Sorry, I’m a bit confused - I’ve only touched one line in the actual config flow. I can see line 232 had the `""` before my PR - Does that need to change to `None` due to the change from `Required` to `Optional`?

DurgNomis-drol replied:
Yes and because you also have changed the tests.

DurgNomis-drol commented on tests/components/motioneye/test_config_flow.py:
```diff
@@ -488,13 +488,13 @@ async def test_advanced_options(hass: HomeAssistant) -> None:
         assert result["data"][CONF_WEBHOOK_SET]
         assert result["data"][CONF_WEBHOOK_SET_OVERWRITE]
-        assert result["data"][CONF_STREAM_URL_TEMPLATE] == "http://moo"
+        assert result["data"][CONF_STREAM_URL_TEMPLATE] == ""
```
```suggestion
        assert result["data"][CONF_STREAM_URL_TEMPLATE] is None
```

We need to test for both. First check if it is `None` and then check if it is string with `isinstance`. You will need to call this function again to change the `CONF_STREAM_URL_TEMPLATE`.

```python
result = await hass.config_entries.options.async_configure(
    result["flow_id"],
    user_input={
        CONF_WEBHOOK_SET: True,         
        CONF_WEBHOOK_SET_OVERWRITE: True,
        CONF_STREAM_URL_TEMPLATE: "http://moo",
    },
)
await hass.async_block_till_done()
```

ZuluWhiskey replied:
Thank you again for your help. I've made the suggested changes to None but can I check...should the code look something like this? Or do I need to do the assertions twice?
```
        result = await hass.config_entries.options.async_configure(
            result["flow_id"],
            user_input={
                CONF_WEBHOOK_SET: True,
                CONF_WEBHOOK_SET_OVERWRITE: True,
                CONF_STREAM_URL_TEMPLATE: None,
            },
        )
        result = await hass.config_entries.options.async_configure(
            result["flow_id"],
            user_input={
                CONF_WEBHOOK_SET: True,
                CONF_WEBHOOK_SET_OVERWRITE: True,
                CONF_STREAM_URL_TEMPLATE: "http://moo",
            },
        )
        await hass.async_block_till_done()
        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
        assert result["data"][CONF_WEBHOOK_SET]
        assert result["data"][CONF_WEBHOOK_SET_OVERWRITE]
        assert result["data"][CONF_STREAM_URL_TEMPLATE] is None or result["data"][CONF_STREAM_URL_TEMPLATE] == "http://moo"
        assert len(mock_setup.mock_calls) == 0
        assert len(mock_setup_entry.mock_calls) == 0
```

ZuluWhiskey replied:
I realised that the above would be useless and two lots of assertations would be needed as the second `result` would overwrite the first `result`. I'm struggling with `isinstance` as I've not used that before - Would the below work instead? First checking the scenario where `CONF_STREAM_URL_TEMPLATE` is `None` and then checking when it is populated with `htto://moo` as was done previously?

```
        result = await hass.config_entries.options.async_configure(
            result["flow_id"],
            user_input={
                CONF_WEBHOOK_SET: True,
                CONF_WEBHOOK_SET_OVERWRITE: True,
                CONF_STREAM_URL_TEMPLATE: None,
            },
        )
        await hass.async_block_till_done()
        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
        assert result["data"][CONF_WEBHOOK_SET]
        assert result["data"][CONF_WEBHOOK_SET_OVERWRITE]
        assert result["data"][CONF_STREAM_URL_TEMPLATE] is None
        assert len(mock_setup.mock_calls) == 0
        assert len(mock_setup_entry.mock_calls) == 0

        result = await hass.config_entries.options.async_configure(
            result["flow_id"],
            user_input={
                CONF_WEBHOOK_SET: True,
                CONF_WEBHOOK_SET_OVERWRITE: True,
                CONF_STREAM_URL_TEMPLATE: "http://moo",
            },
        )
        await hass.async_block_till_done()
        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
        assert result["data"][CONF_WEBHOOK_SET]
        assert result["data"][CONF_WEBHOOK_SET_OVERWRITE]
        assert result["data"][CONF_STREAM_URL_TEMPLATE] == "http://moo"
        assert len(mock_setup.mock_calls) == 0
        assert len(mock_setup_entry.mock_calls) == 0
```

DurgNomis-drol replied:
Yes this will work instead 😃

DurgNomis-drol commented on tests/components/motioneye/test_config_flow.py:
```diff
@@ -488,13 +488,13 @@ async def test_advanced_options(hass: HomeAssistant) -> None:
                 CONF_WEBHOOK_SET: True,
                 CONF_WEBHOOK_SET_OVERWRITE: True,
-                CONF_STREAM_URL_TEMPLATE: "http://moo",
+                CONF_STREAM_URL_TEMPLATE: "",
```
```suggestion
                CONF_STREAM_URL_TEMPLATE: None,

ZuluWhiskey committed:
Update test_config_flow.py

Changing empty string check to None instead, as suggested by DurgNomis-drol

ZuluWhiskey committed:
Add tests for stream url and not

Adding a second test so that both having a stream URL configured or not
are tested

ZuluWhiskey commented:
That should hopefully be resolved now 🤞

DurgNomis-drol commented:
Please see my comment 😃

ZuluWhiskey committed:
Change empty string

Due to the change in this parameter being optional now

balloob committed:
Add new line

ZuluWhiskey commented:
If I'm reading the error correctly, it looks like the test is failing as a string is expected for the stream URL template and we have `None` configured

`voluptuous.error.MultipleInvalid: expected str for dictionary value @ data['stream_url_template']`

Is the fix here to go back to the empty string `""` instead?

balloob commented:
The default value needs to match the validator. The right approach would be to not set a default unless it's in the dictionary. You can do that with `**`. 

```python
stream_kwargs = {}
if CONF_STREAM_URL_TEMPLATE in self._config_entry.options:
    stream_kwargs['default'] = self._config_entry.options[CONF_STREAM_URL_TEMPLATE]

schema[vol.Optional(CONF_STREAM_URL_TEMPLATE, **stream_kwargs)] = str
```

ZuluWhiskey committed:
Update config_flow.py

Don't set a default unless it's in the dictionary

ZuluWhiskey committed:
Merge branch 'patch-1' of git@github.com:ZuluWhiskey/core.git

ZuluWhiskey commented:
> The default value needs to match the validator. The right approach would be to not set a default unless it's in the dictionary. You can do that with `**`. 
> 
> 
> 
> ```python
> 
> stream_kwargs = {}
> 
> if CONF_STREAM_URL_TEMPLATE in self._config_entry.options:
> 
>     stream_kwargs['default'] = self._config_entry.options[CONF_STREAM_URL_TEMPLATE]
> 
> 
> 
> schema[vol.Optional(CONF_STREAM_URL_TEMPLATE, **stream_kwargs)] = str
> 
> ```

Thanks very much for the help Paulus 😃 I've pushed the change and it's working locally for me too 👍

balloob committed:
Fix test

balloob committed:
Change to use suggested value

balloob commented:
Pushed one more change. Can you test @ZuluWhiskey?

This should make sure you can also remove the value.

ZuluWhiskey commented:
> Pushed one more change. Can you test @ZuluWhiskey?
> 
> This should make sure you can also remove the value.

I clearly didn't test fully as you're right, when hitting save it didn't remove the URL template. I've just loaded your push Paulus and it now does remove the URL template - I've tested with it empty, adding a string, and then removing the string - All working
